src/RewardsStreamer.sol

```diff
@@ -123,4 +182,37 @@ contract RewardsStreamer is ReentrancyGuard {
     function getUserInfo(address userAddress) public view returns (UserInfo memory) {
         return users[userAddress];
     }
+
+    /**
+     * @dev Sets the rewards per block for a specified duration. This function can only be called by the owner.
+     * It mints the necessary reward tokens for the specified duration, considering any unassigned rewards.
+     * @param _rewardsPerBlock The new reward rate per block.
+     * @param _durationInBlocks The duration for which the new reward rate should be applied.
+     */
+    function setRewardsPerBlock(uint256 _rewardsPerBlock, uint256 _durationInBlocks) external onlyOwner {
+        require(_durationInBlocks > 0, "Duration must be greater than zero");
+        updateRewardIndex(); // Ensure rewards are up-to-date before changing the rate
+
+        // Calculate the total rewards needed for the specified duration
+        uint256 totalRewardsNeeded = _rewardsPerBlock * _durationInBlocks;
+
+        // Calculate the unassigned rewards currently held by the contract
+        uint256 currentBalance = REWARD_TOKEN.balanceOf(address(this));
+        uint256 unassignedRewards = currentBalance > accountedRewards ? currentBalance - accountedRewards : 0;
+
+        // Calculate the additional rewards needed
+        uint256 additionalRewardsNeeded = 0;
+        if (totalRewardsNeeded > unassignedRewards) {
+            additionalRewardsNeeded = totalRewardsNeeded - unassignedRewards;
+        }
+
+        // Mint the necessary additional reward tokens
+        if (additionalRewardsNeeded > 0) {
+            REWARD_TOKEN.mint(address(this), additionalRewardsNeeded);
+        }
+
+        // Update the rewards per block and reward end block
+        rewardsPerBlock = _rewardsPerBlock;
+        rewardEndBlock = block.number + _durationInBlocks;
+    }
```

Comment:
Okay, so here we're saying, we have an admin that is able to set some reward per block for some duration in blocks.
This allows for issuance of rewards "in the future" because there's a fixed amount of rewards being available for claim per block.

This is opposed to prior this change, where, at any point in time one would update the reward index with whatever rewards are in regardless of how many blocks have passed.

To achieve the same result, one would have to send REWARD_TOKEN to the staking contract on a per block basis, so that `rewardIndex` would be updatable on a per block basis with the newest rewards as well.

I think your proposed change is not bad
What I don't like that much is that the duration unit is blocks. Blocks time may very, so for the same segment of time, there could be different rewards because a different number of blocks has been processed.

I think it'd be more predictable to have a solution that uses actual time as duration instead.

Reply:
> Okay, so here we're saying, we have an admin that is able to set some reward per block for some duration in blocks.

Or you can have epochs and rewards that arrived in the past one determine the current rate of rewards - and you do not need an admin.

> What I don't like that much is that the duration unit is blocks.

Just wanted to show a minimal example and this was the easiest way to do it. In terms of UX, it does make sense to use time.